File: internal/utils/adb.go
package utils

import (
	"fmt"
	"os/exec"
	"strings"
	"time"
)

// TapScreen simulates a tap at the given coordinates
func TapScreen(deviceID, adbPath string, x, y int) error {
	cmd := exec.Command(
		adbPath,
		"-s",
		deviceID,
		"shell",
		"input",
		"tap",
		fmt.Sprintf("%d", x),
		fmt.Sprintf("%d", y),
	)

	return cmd.Run()
}

// SwipeScreen simulates a swipe from (x1, y1) to (x2, y2) with the given duration
func SwipeScreen(deviceID, adbPath string, x1, y1, x2, y2, durationMS int) error {
	cmd := exec.Command(
		adbPath,
		"-s",
		deviceID,
		"shell",
		"input",
		"swipe",
		fmt.Sprintf("%d", x1),
		fmt.Sprintf("%d", y1),
		fmt.Sprintf("%d", x2),
		fmt.Sprintf("%d", y2),
		fmt.Sprintf("%d", durationMS),
	)

	return cmd.Run()
}

// SendText sends text input to the device
func SendText(deviceID, adbPath, text string) error {
	// Replace spaces with %s
	text = strings.Replace(text, " ", "%s", -1)

	cmd := exec.Command(
		adbPath,
		"-s",
		deviceID,
		"shell",
		"input",
		"text",
		text,
	)

	return cmd.Run()
}

// PressKey simulates pressing a key
func PressKey(deviceID, adbPath, keycode string) error {
	cmd := exec.Command(
		adbPath,
		"-s",
		deviceID,
		"shell",
		"input",
		"keyevent",
		keycode,
	)

	return cmd.Run()
}

// IsDeviceConnected checks if the device is connected
func IsDeviceConnected(deviceID, adbPath string) bool {
	cmd := exec.Command(adbPath, "devices")
	output, err := cmd.Output()
	if err != nil {
		return false
	}

	// Check if device ID is in the output
	return strings.Contains(string(output), deviceID)
}

// RestartApp closes and reopens the ROK app
func RestartApp(deviceID, adbPath string) error {
	// Force stop ROK app
	stopCmd := exec.Command(
		adbPath,
		"-s",
		deviceID,
		"shell",
		"am",
		"force-stop",
		"com.lilithgame.roc.gp", // ROK package name
	)
	if err := stopCmd.Run(); err != nil {
		return fmt.Errorf("failed to stop ROK app: %w", err)
	}

	// Wait a bit
	time.Sleep(2 * time.Second)

	// Start ROK app
	startCmd := exec.Command(
		adbPath,
		"-s",
		deviceID,
		"shell",
		"monkey",
		"-p",
		"com.lilithgame.roc.gp",
		"-c",
		"android.intent.category.LAUNCHER",
		"1",
	)
	if err := startCmd.Run(); err != nil {
		return fmt.Errorf("failed to start ROK app: %w", err)
	}

	// Wait for app to load
	time.Sleep(10 * time.Second)

	return nil
}

File: roborok/utils/adb.py
# roborok/utils/adb.py
import subprocess
import logging
from typing import Optional

def tap_screen(device_id: str, adb_path: str, x: int, y: int) -> bool:
    """
    Simulates a tap on the screen
    
    Args:
        device_id: The device ID
        adb_path: Path to the ADB executable
        x: X coordinate
        y: Y coordinate
        
    Returns:
        True if successful, False otherwise
    """
    try:
        subprocess.run(
            [adb_path, "-s", device_id, "shell", "input", "tap", str(x), str(y)],
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        )
        return True
    except subprocess.SubprocessError as e:
        logging.error(f"Error tapping screen at ({x}, {y}): {e}")
        return False

def swipe_screen(device_id: str, adb_path: str, x1: int, y1: int, x2: int, y2: int, duration_ms: int) -> bool:
    """
    Simulates a swipe from (x1, y1) to (x2, y2) with the given duration
    
    Args:
        device_id: The device ID
        adb_path: Path to the ADB executable
        x1: Starting X coordinate
        y1: Starting Y coordinate
        x2: Ending X coordinate
        y2: Ending Y coordinate
        duration_ms: Duration of the swipe in milliseconds
        
    Returns:
        True if successful, False otherwise
    """
    try:
        subprocess.run(
            [
                adb_path, "-s", device_id, "shell", "input", "swipe",
                str(x1), str(y1), str(x2), str(y2), str(duration_ms)
            ],
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE
        )
        return True
    except subprocess.SubprocessError as e:
        logging.error(f"Error swiping screen from ({x1}, {y1}) to ({x2}, {y2}): {e}")
        return False